src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -1379,6 +1383,21 @@ void DatabaseCatalog::dropTableDataTask()
     rescheduleDropTableTask();
 }
 
+void DatabaseCatalog::removeDetachedTableInfo(const TableMarkedAsDropped & table)
+{
+    auto database = tryGetDatabase(table.table_id.getDatabaseName());
+    if (!database)
+        return;
+
+    auto * database_ptr = dynamic_cast<DatabaseOnDisk *>(database.get());
+    if (!database_ptr)
+        return;
+
+    database_ptr->removeDetachedTableInfo(table.table_id);
+
+    database_ptr->DatabaseOnDisk::removeDetachedPermanentlyFlag(getContext(), table.table_id.getNameForLogs(), table.metadata_path, true);
```

Comment:
This is called from `DatabaseCatalog::dropTableFinally`, which is executed asynchronously. Does it mean that the flag remains in the database until the table is finally cleaned up by DatabaseCatalog? What happens if a new table with the same name gets created in the database before the flag is removed? Will the new table get detached permanently on server restart?

Reply:
> This is called from `DatabaseCatalog::dropTableFinally`, which is executed asynchronously. Does it mean that the flag remains in the database until the table is finally cleaned up by DatabaseCatalog? What happens if a new table with the same name gets created in the database before the flag is removed? Will the new table get detached permanently on server restart?

Yes, creating a table with the same name may cause a problem.

I suggest removing the detached flag before deleting the metdata file. 
I assume that if the server restarts between removing the detached flag and deleting the metadata file, it won't be a problem. 
Although we would lose the detached flag, we can recover tables's state manually (perform a detach) and then repeat `DROP DETACHED`.